nntrainer/layers/common_properties.cpp

```diff
@@ -26,6 +28,118 @@ bool Name::isValid(const std::string &v) const {
   return std::regex_match(v, allowed);
 }
 
+ConnectionSpec::ConnectionSpec(const std::vector<props::Name> &layer_ids_,
+                               const std::string &op_type_) :
+  op_type(op_type_),
+  layer_ids(layer_ids_) {
+  NNTR_THROW_IF((op_type != ConnectionSpec::NoneType && layer_ids.size() < 2),
+                std::invalid_argument)
+    << "connection type is not none but has only a single or empty layer id, "
+       "type: "
+    << op_type << " number of names: " << layer_ids.size();
+
+  NNTR_THROW_IF((op_type == ConnectionSpec::NoneType && layer_ids.size() >= 2),
+                std::invalid_argument)
+    << "connection type is none but has only a single or empty layer id, "
+       "number of names: "
+    << layer_ids.size();
+}
+
+ConnectionSpec::ConnectionSpec(const ConnectionSpec &rhs) = default;
+ConnectionSpec &ConnectionSpec::operator=(const ConnectionSpec &rhs) = default;
+ConnectionSpec::ConnectionSpec(ConnectionSpec &&rhs) noexcept = default;
+ConnectionSpec &ConnectionSpec::
+operator=(ConnectionSpec &&rhs) noexcept = default;
+
+bool ConnectionSpec::operator==(const ConnectionSpec &rhs) const {
+  return op_type == rhs.op_type && layer_ids == rhs.layer_ids;
+}
+
+bool InputSpec::isValid(const ConnectionSpec &v) const {
+  return v.getLayerIds().size() > 0;
```

Comment:
is this ok when no input_layer is specified?

Reply:
Next commit #1206 is dealing with the issue.
From #1206,
if a property is not set, it becomes more explicit that it is empty, instead of unsensible default value. So that the property can remain in always valid state or else empty

Reply:
sync.